### PR TITLE
Fix #11572: check_evars_are_solved returns a modified evar map

### LINKED ISSUE #11572: Error: Anomaly "the kernel does not support existential variables."
#### Description of the problem

Found while playing with https://github.com/coq/coq/pull/7825#issuecomment-584784896 (and removing a couple of cases from the final match)
This is a regression of Coq 8.10 and later over Coq 8.9 and earlier.
```coq
(* -*- mode: coq; coq-prog-args: ("-emacs" "-w" "+implicit-core-hint-db,+implicits-in-term,+non-reversible-notation,+deprecated-intros-until-0,+deprecated-focus,+unused-intro-pattern" "-top" "Func" "-R" "." "Top") -*- *)
(* File reduced by coq-bug-finder from original input, then from 466 lines to 305 lines, then from 312 lines to 47 lines, then from 62 lines to 47 lines, then from 383
lines to 47 lines, then from 911 lines to 116 lines, then from 690 lines to 220 lines, then from 717 lines to 374 lines, then from 815 lines to 517 lines, then from 798 lines to 538 lines, then from 820 lines to 539 lines, then from 781 lines to 539 lines, then from 804 lines to 539 lines, then from 776 lines to 537 lines, then from 766 lines to 537 lines, then from 776 lines to 537 lines, then from 768 lines to 537 lines, then from 791 lines to 537 lines, then from 2308 lines to 538 lines, then from 760 lines to 537 lines, then from 760 lines to 537 lines, then from 1798 lines to 538 lines, then from 1294 lines to 537 lines, then from 1998 lines to 538 lines, then from 929 lines to 792 lines, then from 434 lines to 132 lines, then from 412 lines to 364 lines, then from 208 lines to 96 lines, then from 2149 lines to 143 lines, then from 201 lines to 145 lines, then from 161 lines to 145 lines *)
(* coqc version 8.12+alpha (February 2020) compiled on Feb 10 2020 18:42:28 with OCaml 4.06.1
   coqtop version jgross-Leopard-WS:/home/jgross/Documents/repos/coq2,pr/7825 (98d00cacfaf63e66394398a787a044c42a8ab1a3) *)

Global Set Asymmetric Patterns.

Module Export Language.
  Module Export Compilers.
    Module Export type.
      Inductive type {base_type : Type} := base (t : base_type) | arrow (s d : type).
      Global Arguments type : clear implicits.

      Fixpoint final_codomain {base_type} (t : type base_type) : base_type
        := match t with
           | base t
             => t
           | arrow s d => @final_codomain base_type d
           end.

      Fixpoint for_each_lhs_of_arrow {base_type} (f : type base_type -> Type) (t : type base_type) : Type
        := match t with
           | base t => unit
           | arrow s d => f s * @for_each_lhs_of_arrow _ f d
           end.

    End type.
    Notation type := type.type.
    Bind Scope etype_scope with type.type.
    Infix "->" := type.arrow : etype_scope.
    Module base.
      Module Export type.
        Inductive type {base_type : Type} := type_base (t : base_type) | prod (A B : type) | list (A : type) | option (A : type) | unit.
        Global Arguments type : clear implicits.
      End type.
      Notation type := type.type.

    End base.

    Module Export expr.
      Section with_var.
        Context {base_type : Type}.
        Local Notation type := (type base_type).
        Context {ident : type -> Type}
                {var : type -> Type}.

        Inductive expr : type -> Type :=
        | Ident {t} (idc : ident t) : expr t
        | Var {t} (v : var t) : expr t
        | Abs {s d} (f : var s -> expr d) : expr (s -> d)
        | App {s d} (f : expr (s -> d)) (x : expr s) : expr d
        | LetIn {A B} (x : expr A) (f : var A -> expr B) : expr B
        .
      End with_var.

      Module Export APIExports.
        Module base.
          Module Export type.
            Inductive base := Z.
          End type.
          Notation type := (Language.Compilers.base.type APIExports.base.type.base).
        End base.
        Module Import API.
          Notation type := (Language.Compilers.type APIExports.base.type).
          Axiom ident : type -> Type.
          Notation expr := (@expr APIExports.base.type ident).

          Class parameters :=
            {
            }.
          Axiom base_Z : APIExports.base.type.
          Axiom base_listZ : APIExports.base.type.
          Module Export rep.
            Section rep.
              Context {p : parameters}.

              Class rep (t : base.type) :=
                { ltype : Type;
                  rep_p := p }.
              Instance listZ_local {zrep : rep base_Z} : rep base_listZ.
              Admitted.

              Instance Z : rep base_Z.
              Admitted.
            End rep.
          End rep.

          Section defs.
            Context {p : parameters}

                    {listZ : rep.rep base_listZ}.

            Fixpoint base_ltype (t : base.type) : Type :=
              match t with
              | base.type.prod a b => base_ltype a * base_ltype b
              | _ => rep.ltype (rep:=listZ)
              end.
            Fixpoint ltype (t : type.type base.type) : Type :=
              match t with
              | type.base t => base_ltype t
              | type.arrow s d => unit
              end.

          End defs.
          Context {p : parameters}.

          Axiom dummy_base_ltype
            : forall {p : parameters} {listZ : rep base_listZ}
                     (t : base.type), base_ltype t.
          Axiom base_list_lengths : base.type -> Type.
          Axiom cmd : forall {p : parameters}, Type.
          Axiom translate_cmd
            : forall {p : parameters} {t}
                     (e : API.expr (var:=(@ltype p (@listZ_local p (@Z p)))) (type.base t)) (nextn : nat),
              nat * @base_ltype p (@listZ_local p (@Z p)) t * cmd.

          Axiom cmd_skip : forall {p}, @cmd p.

          Fixpoint list_lengths t : Type :=
            match t with
            | type.base t => base_list_lengths t
            | _ => unit
            end.

          Fixpoint translate_func' {t} (e : @API.expr ltype t) (nextn : nat)
            : type.for_each_lhs_of_arrow ltype t ->
              type.for_each_lhs_of_arrow list_lengths t ->
              nat * base_ltype (type.final_codomain t) * cmd :=
            match e with
            | expr.Abs (type.base s) d f =>

              fun (args : base_ltype s * type.for_each_lhs_of_arrow _ d)
                  (lengths : base_list_lengths s * _) =>
                translate_func' (f (fst args)) nextn (snd args) (snd lengths)

            | expr.Ident (type.base b) idc =>
              fun (_ _:unit) => translate_cmd (expr.Ident idc) nextn
            | expr.Var (type.base b) v =>
              fun (_ _:unit) => translate_cmd (expr.Var v) nextn

            | _ => fun _ _ => (0%nat, dummy_base_ltype _, cmd_skip)
            end.
(* Coq 8.10: Error: Anomaly "the kernel does not support existential variables."
Please report at http://coq.inria.fr/bugs/. *)
(* Coq master: Error: Anomaly "grounding a non evar-free term"
Please report at http://coq.inria.fr/bugs/. *)
          (* Coq 8.9, 8.8: Error: Cannot infer the implicit parameter p of cmd whose type is
"parameters" (no type class instance found).
*)
```

#### Coq Version

8.10, master

### PR DESCRIPTION
**Pull request by SkySkimmer**
Even then it's still possible to have unsolved evars after calling it
because of some implicit argument feature I don't understand.

Fixes #11572

### CODE REVIEW

**Review by ejgallego — COMMENTED** [review]
I guess this makes sense but I'm not familiar enough with the spec of the function, so I'd be better if an expert could look at it.

File: pretyping/pretyping.mli [hunk 1]
```diff
@@ -116,10 +116,10 @@ val solve_remaining_evars : ?hook:inference_hook -> inference_flags ->
   env -> ?initial:evar_map -> (* current map *) evar_map -> evar_map
 
 (** Checking evars and pending conversion problems are all solved,
-    reporting an appropriate error message *)
+    reporting an appropriate error message. This call TC inference. *)
```
Comment (JasonGross) [inline pretyping/pretyping.mli]:
```suggestion
    reporting an appropriate error message. This calls TC inference. *)
```

**Comment by JasonGross** [comment]
Does this also fix #11576 ?  If so, there should be a test-suite case

**Comment by ejgallego** [comment]
ccing: @herbelin as to why some parts of the code use `check_evars` and some other do use `check_evars_are_solved` cc #11861

**Comment by herbelin** [comment]
The idea was `check_evars_are_solved` does only `check`. The rôle of solving is the one of `solve_remaining_evars`.

About `check_evars` and `check_evars_are_solved`, the idea (at the time of 78a5b30be750c), was to delay the detection of an error at the latest possible.

There are three ways to complete unification:
- type class search
- `Program` inference_hook
- addressing non-invertible pending problems with heuristics

Before 78a5b30be750c, we possibly had failures arriving too early (in particular because type class search was failing before pending problems were addressed). So, I introduced `check_evars_are_solved` to be used at the very end of the declaration (for `Fixpoint` and `Definition`, and, indirectly, for assumptions and inductive types, via `solve_remaining_evars`).

I have no idea how the three ways to complete unification should ideally be interleaved. This is why I did not change the existing workflow for definitions and fixpoints, and this is why I did not use `solve_remaining_evars` in definitions and fixpoints (for instance I kept the call to TC where it was, only doing that it does not fail too early).

Basically, the conclusion of all this is that, for each declaration, the unification workflow should basically be:

- do all invertible unification simplication
- apply heuristics (in some order...)
- check that all evars are solved

Morally, the first one is a succession of calls to `evar_conv` via `interp_constr`, `interp_type`, `interp_context`, even though `evar_conv` includes in practice heuristics applied too early along the way 😒.

Morally, the second one is `solve_remaining_evars` (but in practice, definitions and fixpoints do their own sauce of mixing TC and heuristics, so they don't use `solve_remaining_evars` explicitly).

The last one is `check_evars_are_solved`.

What is otherwise the status of `check_evars`? I believe it should be replaced by `check_evars_are_solved` to get better error messages and to correctly check that postponed problems are solved. In `record.ml`, and `hints.ml`, and `classes.ml`, and `Context`, there is no reason to not check also the postponed problems. I have just been too incremental in 78a5b30be750c.

TL;DR:

- no reason to change the rôle `check_evars_are_solved` (imo) and investigate the cause of the #11572 regression
- replace `check_evars` by `check_evars_are_solved` everywhere
- longer term: understand if there is a clear picture at how TC and postponed constraints should interact
- understand if there is a clever way to fail at unresolved TC evars than replaying TC with the fail flag

**Comment by herbelin** [comment]
@skyskimmer: I think it is worth to make a pass on the use of `check_evars` and `check_evars_are_solved` as I outlined above. Don't know what are your intentions.

**Comment by SkySkimmer** [comment]
I actually don't like `check_evars_are_solved`, IMO if some evar which doesn't appear in the final term remains unsolved we should allow finishing the definition.
For instance
~~~coq
Definition foo : True.
Proof.
  let x := open_constr:(_) in exact I.
Defined.
~~~
is accepted (with a warning) but using `:= ltac:()` fails, and I prefer accepting.

**Comment by herbelin** [comment]
> I actually don't like `check_evars_are_solved`, IMO if some evar which doesn't appear in the final term remains unsolved we should allow finishing the definition.
> For instance
> 
> ```coq
> Definition foo : True.
> Proof.
>   let x := open_constr:(_) in exact I.
> Defined.
> ```
> 
> is accepted (with a warning) but using `:= ltac:()` fails, and I prefer accepting.

Why not accepting it indeed. So, you would instead like a `check_used_evars_are_solved` then?

**Comment by SkySkimmer** [comment]
>Why not accepting it indeed. So, you would instead like a check_used_evars_are_solved then?

Isn't that just `check_evars`?

**Comment by herbelin** [comment]
> Isn't that just `check_evars`?

`check_evars` does not check for unresolved postponed problems, right? But, of course, this is also a question of whether some other function (e.g. `solve_unif_constraints_with_heuristics` or `apply_heuristics` with the right flags) have already been applied to check postponed problems.

If we want to tolerate remaining non-occurring evars, as well as postponed problems about these evars, some work is needed.

**Comment by SkySkimmer** [comment]
>unresolved postponed problems

I forgot those
Not sure what I want for them

**Comment by Zimmi48** [comment]
So, what should happen to this PR? Should it get closed?

**Comment by maximedenes** [comment]
@SkySkimmer if this is not ready, can we turn it into a draft?

**Comment by ppedrot** [comment]
@SkySkimmer ping?

**Comment by maximedenes** [comment]
@herbelin I seem to need this patch for #7825 to work properly. I re-read the discussion, there's still a basic point I don't understand : what is the point of calling TC inference and throwing away the solution ? (which is what `check_evars_are_solved` does)

Said otherwise, when `check_evars_are_solved` succeeds, what can we say about the term and its evarmap?

**Comment by maximedenes** [comment]
> Said otherwise, when `check_evars_are_solved` succeeds, what can we say about the term and its evarmap?

In fact, I guess I imagine the intention: `check_evars_are_solved` should not perform any work (i.e. not instantiate stuff), beyond just reaching the failing path and backtracking.

**Comment by mattam82** [comment]
My viewpoint is that check_evars_are_solved should return unit (and error on undefined, unresolvable typeclass evars) and maybe solve_remaining_evars should be called for definition and fixpoint (e.g. before check_evars_are_solved in Declare.prepare_definition).

**Comment by mattam82** [comment]
We might improve the way to get a meaningful unresolved typeclass constraint error message by changing the API of typeclasses and adding a `check_typeclasses_are_solved` in class_tactics.ml

**Comment by SkySkimmer** [comment]
https://github.com/coq/coq/pull/12920 is better
